File: src/main/java/com/sf_lolitahag/hanabi/Firework.java
/*
Copyright (C) 2016 Koji Wakamiya
Copyright (C) 2016 Masanori Kojima
Released under the MIT license
http://opensource.org/licenses/mit-license.php
*/
package com.sf_lolitahag.hanabi;

import com.sf_lolitahag.Utils;
import java.awt.Graphics;
import java.util.ArrayList;
import javax.swing.Timer;

public class Firework {

  private static final int POSITION_BASE = 400;
  private static final int POSITION_GAP = 300;
  private static final int TIMER_POSITION_UPDATE_FIREBALL = 25;
  private static final int TIMER_POSITION_UPDATE_SPARK = 150;
  private static final int SPARK_COUNT_MIN = 50;
  private static final int SPARK_COUNT_MAX = 250;
  private static final int FIREBALL_LIFETIME_BASE = 3000;
  private static final int FIREBALL_LIFETIME_COEFFICIENT = 2500;
  private static final int SPARK_LIFETIME_BASE = 1250;
  private static final int SPARK_LIFETIME_COEFFICIENT = 1000;
  private static final double LAUNCH_PERCENT = 0.7;
  private boolean isRun = false;
  private boolean isFireballShow = false;
  private boolean isSparksShow = false;
  private final Fireball fireball = new Fireball();
  private final ArrayList<Spark> sparks = new ArrayList<>();
  private final Timer updateTimer = new Timer(TIMER_POSITION_UPDATE_FIREBALL,
      (e) -> updatePosition());
  private Timer fireballTimer;
  private Timer sparkTimer;

  public Firework() {
  }

  public void draw(final Graphics g) {
    if (isFireballShow) {
      fireball.draw(g);
    } else if (isSparksShow) {
      sparks.parallelStream().forEach(spark -> spark.draw(g));
    }
  }

  public void startFireball() {
    if (!isRun && Math.random() > LAUNCH_PERCENT) {
      isRun = true;

      initFireball(Utils.getRandBaseCoe(POSITION_BASE, POSITION_GAP), POSITION_BASE);
      isFireballShow = true;
      fireballTimer = new Timer(
          Utils.getRandBaseCoe(FIREBALL_LIFETIME_BASE, FIREBALL_LIFETIME_COEFFICIENT),
          (e) -> onFinishFireball());
      fireballTimer.start();

      updateTimer.setDelay(TIMER_POSITION_UPDATE_FIREBALL);
      updateTimer.start();
    }
  }

  private void startSpark() {
    initSpark(fireball.getTopX(), fireball.getTopY());
    isSparksShow = true;
    sparkTimer = new Timer(Utils.getRandBaseCoe(SPARK_LIFETIME_BASE, SPARK_LIFETIME_COEFFICIENT),
        (e) -> onFinishSpark());
    sparkTimer.start();

    updateTimer.setDelay(TIMER_POSITION_UPDATE_SPARK);
    updateTimer.restart();
  }

  private void initFireball(int x, int y) {
    fireball.init(x, y);
  }

  private void initSpark(int x, int y) {
    sparks.clear();
    int sparkCount = Utils.getRandRange(SPARK_COUNT_MIN, SPARK_COUNT_MAX);
    for (int index = 0; index < sparkCount; index++) {
      sparks.add(new Spark());
    }
    sparks.forEach(spark -> spark.init(x, y));
  }

  private void updatePosition() {
    if (isFireballShow) {
      fireball.next();
    } else if (isSparksShow) {
      sparks.forEach(Spark::next);
    }
  }

  private void onFinishFireball() {
    updateTimer.stop();

    isFireballShow = false;
    fireballTimer.stop();

    startSpark();
  }

  private void onFinishSpark() {
    isSparksShow = false;
    sparkTimer.stop();

    updateTimer.stop();
    isRun = false;
  }
}

File: src/main/kotlin/com/sf_lolitahag/hanabi/Firework.kt
/*
Copyright (C) 2016 Koji Wakamiya
Copyright (C) 2016 Masanori Kojima
Released under the MIT license
http://opensource.org/licenses/mit-license.php
*/
package com.sf_lolitahag.hanabi

import com.sf_lolitahag.Utils
import java.awt.Graphics
import java.util.*
import javax.swing.Timer

class Firework {
    private var isRun = false
    private var isFireballShow = false
    private var isSparksShow = false
    private val fireball = Fireball()
    private val sparks = ArrayList<Spark>()
    private val updateTimer = Timer(TIMER_POSITION_UPDATE_FIREBALL
    ) { e -> updatePosition() }
    private var fireballTimer: Timer? = null
    private var sparkTimer: Timer? = null

    val isNotRun: Boolean
        get() = !isRun

    fun draw(g: Graphics) {
        if (isFireballShow) {
            fireball.draw(g)
        } else if (isSparksShow) {
            sparks.parallelStream().forEach { spark -> spark.draw(g) }
        }
    }

    fun startFireball() {
        if (Math.random() > LAUNCH_PERCENT) {
            isRun = true

            initFireball(Utils.getRandBaseCoe(POSITION_BASE, POSITION_GAP), POSITION_BASE)
            fireballTimer = Timer(Utils.getRandBaseCoe(FIREBALL_LIFETIME_BASE, FIREBALL_LIFETIME_COEFFICIENT)
            ) { e -> onFinishFireball() }
            fireballTimer!!.start()
            isFireballShow = true

            updateTimer.delay = TIMER_POSITION_UPDATE_FIREBALL
            updateTimer.start()
        }
    }

    private fun startSpark() {
        initSpark(fireball.topX, fireball.topY)
        isSparksShow = true
        sparkTimer = Timer(Utils.getRandBaseCoe(SPARK_LIFETIME_BASE, SPARK_LIFETIME_COEFFICIENT)
        ) { e -> onFinishSpark() }
        sparkTimer!!.start()

        updateTimer.delay = TIMER_POSITION_UPDATE_SPARK
        updateTimer.restart()
    }

    private fun initFireball(x: Int, y: Int) {
        fireball.init(x, y)
    }

    private fun initSpark(x: Int, y: Int) {
        sparks.clear()
        val sparkCount = Utils.getRandRange(SPARK_COUNT_MIN, SPARK_COUNT_MAX)
        for (index in 0 until sparkCount) {
            sparks.add(Spark())
        }
        sparks.forEach { spark -> spark.init(x, y) }
    }

    private fun updatePosition() {
        if (isFireballShow) {
            fireball.next()
        } else if (isSparksShow) {
            sparks.forEach { it.next() }
        }
    }

    private fun onFinishFireball() {
        updateTimer.stop()

        isFireballShow = false
        fireballTimer!!.stop()

        startSpark()
    }

    private fun onFinishSpark() {
        isSparksShow = false
        sparkTimer!!.stop()

        updateTimer.stop()
        isRun = false
    }

    companion object {

        private const val POSITION_BASE = 400
        private const val POSITION_GAP = 300
        private const val TIMER_POSITION_UPDATE_FIREBALL = 30
        private const val TIMER_POSITION_UPDATE_SPARK = 150
        private const val SPARK_COUNT_MIN = 50
        private const val SPARK_COUNT_MAX = 250
        private const val FIREBALL_LIFETIME_BASE = 3000
        private const val FIREBALL_LIFETIME_COEFFICIENT = 2500
        private const val SPARK_LIFETIME_BASE = 1250
        private const val SPARK_LIFETIME_COEFFICIENT = 1000
        private const val LAUNCH_PERCENT = 0.85
    }
}
